core/src/main/scala/com/microsoft/azure/synapse/ml/logging/Usage/TokenUtils.scala

```diff
@@ -0,0 +1,156 @@
+import spray.json.DefaultJsonProtocol.{StringJsonFormat, jsonFormat3}
+
+import java.util.UUID
+import com.microsoft.azure.synapse.ml.logging.common.WebUtils._
+
+import java.util.Date
+import pdi.jwt._
+import org.json.JSONObject
+import spray.json.RootJsonFormat
+
+import scala.util.{Failure, Success, Try}
+
+case class MwcToken (TargetUriHost: String, CapacityObjectId: String, Token: String)
+object TokenUtils {
+  var AADToken: String = ""
+  val MwcWorkloadTypeMl = "ML"
+
+  def getAccessToken(): String = {
+    if (checkTokenValid(this.AADToken))
+      this.AADToken
+    else {
+      refreshAccessToken()
+      this.AADToken
+    }
+  }
+
+  def getAccessToken(tokenType: String): String = {
+
+    val objectName = "com.microsoft.azure.trident.tokenlibrary.TokenLibrary"
+    val mirror = currentMirror
+    val module = mirror.staticModule(objectName)
+    val obj = mirror.reflectModule(module).instance
+    val objType = mirror.reflect(obj).symbol.toType
+    val methodName = "getAccessToken"
+    val methodSymbols = objType.decl(TermName(methodName)).asTerm.alternatives
+    val argType = typeOf[String]
+
+    val selectedMethodSymbol = methodSymbols.find { m =>
+      m.asMethod.paramLists match {
+        case List(List(param)) => param.typeSignature =:= argType
+        case _ => false
+      }
+    }.getOrElse(throw new NoSuchMethodException(s"Method $methodName with argument type $argType not found"))
+
+    val methodMirror = mirror.reflect(obj).reflectMethod(selectedMethodSymbol.asMethod)
+    methodMirror(tokenType).asInstanceOf[String]
+  }
+
+  private def checkTokenValid(token: String): Boolean = {
+    try{
+      val expiryDate: Date = getExpiry(token)
+      val expiryEpoch = expiryDate.toInstant.getEpochSecond
+      val now = Instant.now().getEpochSecond
+      now < expiryEpoch - 60
+    }
+    catch
+    {
+      case e: Exception =>
+        SynapseMLLogging.logMessage(s"TokenUtils::checkTokValid: Token {$token} parsing went wrong (usage test). " +
+          s"Exception = $e")
+        false
+    }
+    /*if (token == null || token.isEmpty()) {
+      false
+    }
+    try {
+      val parsedToken = token.parseJson.asJsObject()
+      val expTime = parsedToken.fields("exp").convertTo[Int]
+      val now = Instant.now().getEpochSecond()
+      now < expTime - 60
+    } catch {
+      case e: Exception => {
+        SynapseMLLogging.logMessage(s"TokenUtils::checkTokValid: Token {$token} parsing went wrong (usage test).")
+        false
+      }
+    }*/
+  }
+
+  private def refreshAccessToken(): Unit = {
+    try {
+      if (SparkContext.getOrCreate() != null) {
+        val token = getAccessToken("pbi")
+        AADToken = token
+      } else {
+        val token = new FabricTokenServiceClient().getAccessToken("pbi")
+        AADToken = token
+      }
+    } catch {
+      case e: Exception =>
+        SynapseMLLogging.logMessage(s"refreshAccessTok: failed to refresh pbi tok. Exception: {$e}. (usage test)")
+    }
```

Comment:
Try not to make blanket exception catches, let the exceptions percolate up in general. If you need to stop an exception, ask yourself if something is wrong in the logic that is causing the exception. If its truly unavoidable, make the exception as specific as possible to avoid hiding bugs

Reply:
ditto: As prior comment